backend/main/shared/dps.go

```diff
@@ -0,0 +1,213 @@
+		return &adapter
+	}
+
+	// Initialize the DPS API client.
+	codec := zbor.NewCodec()
+
+	conn, err := grpc.DialContext(
+		adapter.Context,
+		adapter.URL,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.FailOnNonTempDialError(true), // DialContext will fail immediately if passed bad connection string
+		grpc.WithBlock(),                  // connection should be blocking
+	)
+
+	if err != nil {
+		log.Error().Str("dps", adapter.URL).Err(err).Msg("could not dial API host")
+		panic("error initializing DPS gRPC connection")
+	}
+
+	client := dpsApi.NewAPIClient(conn)
+	index := dpsApi.IndexFromAPI(client, codec)
+
+	cacheSize := uint64(100_000_000)
+	_invoker, err := invoker.New(index, invoker.WithCacheSize(cacheSize))
+	adapter.Invoker = _invoker
+
+	if err != nil {
+		log.Error().Str("dps", adapter.URL).Err(err)
+		panic("error initializing DPS Invoker")
+	}
+	return &adapter
+}
+
+// FLOW token has its own function for fetching balance because
+// it has its own script to account for locked/staked tokens.
+func (dps *DpsAdapter) GetFlowBalanceAtBlockheightScript(addr string, blockheight uint64) (cadence.Value, error) {
+	var args []cadence.Value
+
+	// Prepare Args
+	flowAddress := flow.HexToAddress(addr)
+	cadenceAddress := cadence.NewAddress(flowAddress)
+
+	args = append(args, cadenceAddress)
+
+	_script, err := ioutil.ReadFile("./main/cadence/scripts/get_flow_token_balances.cdc")
+
+	if err != nil {
+		log.Error().Err(err).Msgf("Error reading cadence script file.")
+		return nil, err
+	}
+
+	// Replace script args
+	script := dps.Config.InsertCoreContractAddresses(string(_script))
+
+	result, err := dps.Invoker.Script(blockheight, []byte(script), args)
+	if err != nil {
+		log.Error().Err(err).Msg("invoker.Script failed")
+		return nil, err
+	}
+	return result, nil
+}
+
+func (dps *DpsAdapter) GetBalanceAtBlockheight(addr string, blockheight uint64, c *Contract) (*FTBalanceResponse, error) {
+	if dps.bypass() {
+		log.Info().Msgf("overriding dps service for emulator")
+		return &DummyBalance, nil
+	}
+
+	var err error
+	var result cadence.Value
+
+	tokenBalanceResponse := FTBalanceResponse{}
+	tokenBalanceResponse.NewFTBalance()
+	tokenBalanceResponse.Addr = addr
+	tokenBalanceResponse.BlockHeight = blockheight
+
+	if *c.Name == "FlowToken" {
+		result, err = dps.GetFlowBalanceAtBlockheightScript(addr, blockheight)
+		if err != nil {
+			return nil, err
+		}
+		fields := result.(cadence.Struct).Fields
+		tokenBalanceResponse.PrimaryAccountBalance = fields[0].ToGoValue().(uint64)
+		tokenBalanceResponse.StakingBalance = fields[3].ToGoValue().(uint64)
+		tokenBalanceResponse.Stakes = strings.Split(fields[5].ToGoValue().(string), ", ")
+
+		return &tokenBalanceResponse, nil
+	} else {
+		result, err = dps.GetTokenBalanceAtBlockheightScript(addr, blockheight, c)
+		if err != nil {
+			return nil, err
+		}
+		fields := result.(cadence.Struct).Fields
+		tokenBalanceResponse.Balance = uint64(fields[1].ToGoValue().(uint64))
+
```

Comment:
lets hope the index number of these values returned doesn't change ... seems strange to me that Non-Flow FT's are at index 0 and `FlowToken` `PrimaryAccountBalance` is at index 1.`

Reply:
What happens when an account has no balance? does it return with `0` or `nil`?

Reply:
`0` -- the `AccountInfo` struct initializes all values and then updates relevant ones after it fetches:

```
pub struct AccountInfo {
    pub(set) var primaryAcctBalance: UFix64
    pub(set) var secondaryAddress: Address?
    pub(set) var secondaryAcctBalance: UFix64
    pub(set) var stakedBalance: UFix64
    pub(set) var hasVault: Bool
    pub(set) var stakes: String

    init() {
        self.primaryAcctBalance = 0.0 as UFix64
        self.secondaryAddress = nil
        self.secondaryAcctBalance = 0.0 as UFix64
        self.stakedBalance = 0.0 as UFix64
        self.hasVault = true
        self.stakes = ""
    }
}
```

Reply:
thanks for sharing this info. 0 is good.